openweathermap-ios/openweathermap-ios/Weather/Model/WeatherManager.swift

```diff
@@ -14,31 +14,31 @@ struct WeatherManager {
         getRequest(with: urlString)
     }
     
-    func getRequest(with utlString: String) {
-        if let url = URL(string: utlString) {
-            let session = URLSession(configuration: .default)
-            let task = session.dataTask(with: url) { (data, response, erro) in
-                if erro != nil{
-                    return
-                }
-                if let safeData = data {
-                    if let weather = self.parseJSON(safeData){
-                        self.delegate?.didUpdateWeather(self, weather: weather)
-                    }
+    private func getRequest(with utlString: String) {
+        guard let url = URL(string: utlString) else { return }
+        let session = URLSession(configuration: .default)
+        let task = session.dataTask(with: url) { (data, response, erro) in
+            if erro != nil{
+                return
+            }
+            if let safeData = data {
+                if let weather = self.parseJSON(safeData){
+                    self.delegate?.didUpdateWeather(weather: weather)
                 }
             }
-            task.resume()
         }
+        task.resume()
     }
     
-    func parseJSON(_ weatherData: Data) -> WeatherModel? {
+    private func parseJSON(_ weatherData: Data) -> WeatherDataResponse? {
         let decoder = JSONDecoder()
         do {
             let decodeData = try decoder.decode(WeatherData.self, from: weatherData)
-            let name = decodeData.name
+            let cityName = decodeData.name
             let temp = decodeData.main.temp
             let description = decodeData.weather[0].description
-            let weather = WeatherModel(city: name, temp: temp, description: description)
+            let id = decodeData.weather[0].id
```

Comment:
Is it safe to access the array this way?

Reply:
Yes, this is due to the guard let safeData  that checks the data conditional.

`guard let safeData = data else { return }`

Reply:
I don't agree with you. Accessing arrays by index without any check above is always a risk of a crash. 

Your `guard let` is validating only the existence of data but does not guarantee the existence of items inside the property weather. Unless the API says something about it. (but even with the API saying something, I prefer to go safe, always)

Reply:
But how to check the data since I tried to use if let and guard let and xcode shows this message "Initializer for conditional binding must have Optional type, not 'Int'"
What to do?

Reply:
This is a UX decision. What do YOU want?

A) show empty values to the user;
B) show feedback error to the user;
C) or crash

Reply:
Is A or B good practice?
Regardless of the solution, I don't know how I could solve it. What should I study?

Reply:
My personal preference is to treat it as an error. Like if the server got response something wrong. That said, I would implement a failable initializer and let the rest of the code handle that.

https://developer.apple.com/swift/blog/?id=17

Reply:
The idea is to check if the data exists, but it is not doing as expected.
When trying to verify that the data is whole, I also create an option. You have to check that out.
I go into a loop that doesn't solve anything.
Is there the way or what I'm doing is nothing to do :).

```
guard let id = Int(fromInt: decodeData.weather[0].id) else { return nil}
```
```
extension Int {
    init?(fromInt: Int) {
        if fromInt == Int() {
            self = fromInt
        } else {
            return nil
        }
    }
}
```

Reply:
Forget what I said about failable initializer. I just notice that the logic's scope is a function and not a constructor. 

So just add a `guard let weather = weather.first else { return nil }` and done. Your method is already optional, so it is enough for now.

Reply:
And to not create a conflict with the next constant, consider return directly what you want:

```swift
return WeatherDataResponse(
    city: cityName, 
    temp: temp, 
    description: description, conditionId: id
)
```

Reply:
Doesn't a return need a function?

Reply:
What I tried to say is replace this:
```swift
let weather = WeatherDataResponse(city: cityName, temp: temp, description: description, conditionId: id)
return weather
```

for this:
```swift
return WeatherDataResponse(city: cityName, temp: temp, description: description, conditionId: id)
```

Reply:
You don't need the attribution if you gonna return immediately.

Reply:
I didn't understand :).

Reply:
I am talking about the end of the method, have you noticed?

I just follow the same thread because I was trying to avoid the variable name's conflicts.